### PR TITLE
Examples visual

### PR DESCRIPTION
**Pull request by tias**
Reworked visual examples: as notebooks, and MUCH simpler visualisations with the usual visualisation tools (pandas, plotly)

@IgnaceBleukx can you do a code review? I tried to follow the style that we discussed for the csplib examples. Also, a somewhat different way of checking the result of model.solve() while still being able to work at the 'top level' of a cell...

Do you know of other examples that have a visualisation that we could add?

Maybe VRP/TSP? we have 'some' visualisation in advanced/VRP by learning...ipynb

### CODE REVIEW

File: examples/visual/car_sequencing.ipynb [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+   "source": [
+    "# the CPMpy model and variables\n",
+    "def model_car_sequence(demand, per_slots, at_most, requires):\n",
+    "    nSlots = sum(demand)  # The amount of timeslots to be filled\n",
+    "    nOptions = len(at_most)  # The amount of different options\n",
+    "    nCarConfigs = len(demand)  # The amount of different car types\n",
+    "    requires = cpm_array(requires)  # For element constraint\n",
+    "\n",
+    "    # Decision variables\n",
+    "    slots = intvar(1, nCarConfigs, shape=nSlots, name=\"slots\")  # Sequence of different car types\n",
```
Comment (IgnaceBleukx) [inline examples/visual/car_sequencing.ipynb:57]: Writing it 0-indexed might be a little clearer in the constraint themselves? But 1-indexed seems the way it is done most of the time indeed....

File: examples/visual/car_sequencing.ipynb [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    "    m += [sum(slots == (i+1)) == demand[i] for i in range(nCarConfigs)]\n",
+    "\n",
+    "    # Make sure that the options in the setup table correspond to those of the car type\n",
+    "    for s in range(nSlots):\n",
+    "        m += [setup[s,o] == requires[slots[s]-1,o] for o in range(nOptions)]\n",
+    "\n",
+    "    # Check that no more than \"at most\" car properties are used per \"per_slots\" slots\n",
+    "    for o in range(nOptions):\n",
+    "        for s in range(nSlots - per_slots[o]):\n",
+    "            slotrange = range(s, s + per_slots[o])\n",
```
Comment (IgnaceBleukx) [inline examples/visual/car_sequencing.ipynb:73]: In the csp-lib version I used the `rolling_window` function from numpy. For me it's more clear using that one but this is fine too

File: examples/visual/packing_rectangles.ipynb [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    "\n",
+    "    m = Model()\n",
+    "\n",
+    "    ## Necessary constraints\n",
+    "    # Every item has to be within the overall area\n",
+    "    m += [pos_x + widths <= total_x,\n",
+    "          pos_y + heights <= total_y]\n",
+    "\n",
+    "    # No-overlap: every item has to be fully above, below or next to every other item\n",
+    "    for i in range(n):\n",
```
Comment (IgnaceBleukx) [inline examples/visual/packing_rectangles.ipynb:67]: Why not use the `all_pairs` helper function?

**Review by IgnaceBleukx — COMMENTED** [review]
Diff of scheduling.ipynb was too large for github to review:
- For the precedences, I like to use `start[:,-1] <= end[:,1:]`
- Same for the optional constraint of job B
- I would also use snake_case in this file

**Comment by IgnaceBleukx** [comment]
We could also add a nonogram example? Depending on the picture, it could be really nice

**Comment by IgnaceBleukx** [comment]
Other problems on the wishlist:
- RCPSP
- Nonogram (see examples/csplib/prob012_nonogram.py) which can be made cleaner after merging #121

File: examples/visual/car_sequencing.ipynb [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+   "id": "included-shadow",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# the visualisation: build dataframe, return/display it (is automatically nicely)\n",
+    "def visualize_car_sequence(slots, setup):\n",
+    "    df = pd.DataFrame(setup).astype(str).replace('True', 'X').replace('False', '')\n",
+    "    df['Config (slots)'] = [f\"Options {l}\" for l in slots]  # insert as first column\n",
+    "    df = df.set_index('Config (slots)')\n",
+    "    return df"
```
[the remark this answers is not in the document]
Comment (tias) [reply examples/visual/car_sequencing.ipynb:91]: Yes, do you have an example/snippet?

**Comment by tias** [comment]
Addressed the earlier comments.

Added visual tables for nqueens and sudoku, pretty cute.

Moved all visual examples to the top-level examples folder, and made a README that lists the notebooks and what they are examples of.

Ideally there would be colab/binder/github links for each of these, but I did not have the patience for that.


I will merge this branch by the end of the week so the CSAI students can see/use them. Any comments/additions welcome till then.

**Pull request merged**